File: classlink-oneroster-oauth1/php/OneRoster.php
<?php

// Uses curl
class OneRoster {
    private $client_id;
    private $client_secret;

    function __construct($client_id, $client_secret) {
        $this->client_id = $client_id;
        $this->client_secret = $client_secret;
    }

    /**
     * Makes a get request to the given url with the stored key and secret
     *
     * @param string $url      The exact url of the request, including all params
     * @return array           Status_code and response from the request
     */
    public function makeRosterRequest($url) {

        // Create the timestamp and nonce
        $timestamp = strval(time());
        $nonce = $this->generateNonce(strlen($timestamp));

        // Assign the oauth params
        $oauth = array( 'oauth_consumer_key' => $this->client_id,
            'oauth_signature_method' => 'HMAC-SHA256',
            'oauth_timestamp' => $timestamp,
            'oauth_nonce' => $nonce);

        // Split the url into the base url and the params
        $url_pieces = explode("?", $url);
        $params = $oauth;

        // Add the url params if they exist
        if (count($url_pieces) == 2) {
            $url_params = $this->paramsToArray($url_pieces[1]);
            $params = array_merge($params, $url_params);
        }

        // Generate the oauth signature
        $base_info = $this->buildBaseString($url_pieces[0], 'GET', $params);
        $composite_key = rawurlencode($this->client_secret) . '&';
        $oauth_signature = base64_encode(hash_hmac('SHA256', $base_info, $composite_key, true));
        $oauth['oauth_signature'] = $oauth_signature;

        // Create the oauth header
        $auth_header = $this->buildAuthorizationHeader($oauth);

        // Return the response from the get request
        return $this->makeCurlRequest($auth_header, $url);
    }

    /**
     * Generates the base string for the generation of the oauth signature
     *
     * @param string $baseURI       The base url
     * @param string $method        The HTTP method
     * @param array $params         The url params
     * @return string               The base string used to generate the signature
     *
     */
    private function buildBaseString($baseURI, $method, $params){
        $r = array();
        ksort($params);
        foreach($params as $key=>$value){
            $r[] = "$key=" . rawurlencode($value);
        }
        return $method."&" . rawurlencode($baseURI) . '&' . rawurlencode(implode('&', $r));
    }

    /**
     * Generates the nonce
     *
     * @param int $length       Length of the nonce
     * @return string           Nonce
     */
    private function generateNonce($length) {
        $characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
        $charactersLength = strlen($characters);
        $randomString = '';
        for ($i = 0; $i < $length; $i++) {
            $randomString .= $characters[rand(0, $charactersLength - 1)];
        }
        return $randomString;
    }

    /**
     * Generates the authorization header with the auth info
     *
     * @param array $oauthinfo     The parameters for the oauth header
     * @return string               The oauth header
     */
    private function buildAuthorizationHeader($oauthinfo)
    {
        $r = 'Authorization: OAuth ';
        $values = array();
        foreach ($oauthinfo as $key => $value)
            $values[] = "$key=\"" . rawurlencode($value) . "\"";

        $r .= implode(',', $values);
        return $r;
    }

    /**
     * Converts the url params into an array
     *
     * @param string $url_params    The params from the url
     * @return array                An array of the params
     */
    private function paramsToArray($url_params) {
        $params = explode("&", $url_params);
        $result = array();
        foreach ($params as $value) {
            $value = rawurldecode($value);
            $split = explode("=", $value);
            if (count($split) == 2) {
                $result[$split[0]] = $split[1];
            } else {
                $result["filter"] = substr($value, 7);
            }

        }
        return $result;
    }

    /**
     * Performs the desired get request and returns an array with the status_code and response body
     *
     * @param string $auth_header   The auth header
     * @param string $url           The request url
     * @return array                The staus_code and response
     */
    private function makeCurlRequest($auth_header, $url) {
        $curl = curl_init();

        curl_setopt_array($curl, array(
            CURLOPT_URL => str_replace(" ", "%20", $url),
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => "",
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_CUSTOMREQUEST => "GET",
            CURLOPT_HTTPHEADER => array(
                $auth_header,
            ),
        ));

        $response = curl_exec($curl);
        $httpcode = curl_getinfo($curl, CURLINFO_HTTP_CODE);

        curl_close($curl);

        if ($httpcode == 0) {
            return array("status_code" => $httpcode,
                "response" => "An error occurred, check your URL");
        }

        return array("status_code" => $httpcode,
                    "response" => $response);
    }
}

File: classlink-oneroster-oauth1/go/OneRoster/OneRoster.go
package OneRoster

import (
	"io/ioutil"
	"net/http"
	"strings"
	"time"
	"fmt"
	"math/rand"
	"net/url"
	"sort"
	"encoding/base64"
	"crypto/hmac"
	"crypto/sha256"
)

// Object for the OneRoster containing client_id and client_secret
type OneRoster struct {
	clientId string
	clientSecret string
}

// Constructor for OneRoster
func New(clientId string, clientSecret string) OneRoster {
	rr := OneRoster {clientId, clientSecret}
	return rr
}


// Makes the request to the given OneRoster URL with the stored key and secret
// It returns a the status code and JSON response
func (rr OneRoster) MakeRosterRequest(reqUrl string) (int, string) {
	timestamp := fmt.Sprint(time.Now().Unix())
	nonce := generateNonce(len(timestamp))

	oauth := make(map[string]string)
	oauth["oauth_consumer_key"] = rr.clientId
	oauth["oauth_signature_method"] = "HMAC-SHA256"
	oauth["oauth_timestamp"] = timestamp
	oauth["oauth_nonce"] = nonce

	urlPieces := strings.Split(reqUrl, "?")
	allParams := make(map[string]string)
	if len(urlPieces) == 2 {
		urlParams := paramsToMap(urlPieces[1])
		allParams = mergeParams(urlParams, oauth)
	} else {
		allParams = oauth
	}
	baseInfo := buildBaseString(urlPieces[0], "GET", allParams)
	compositeKey := url.QueryEscape(rr.clientSecret) + "&"
	
	oauth["oauth_signature"] = generateSig(baseInfo, compositeKey)

	authHeader := buildAuthHeader(oauth)
	return makeGetRequest(reqUrl, authHeader)
}

// Makes the actual request to the URL with the generated auth header
func makeGetRequest(reqUrl, authHeader string) (int, string) {
	hc := http.Client{}
	req, _ := http.NewRequest("GET", reqUrl, nil)

	req.Header.Add("Authorization", authHeader)
	resp, err := hc.Do(req)
	if (err != nil) {
		return 0, "An error occurred, check your URL"
	}
	bodyBytes, _ := ioutil.ReadAll(resp.Body)
	return resp.StatusCode, string(bodyBytes)
}

// Creates the auth header from a map of the oauth parameters
func buildAuthHeader(oauthInfo map[string]string) string {
	result := "OAuth "
	values := make([]string, 0)
	for key, value := range oauthInfo {
		values = append(values, key + "=\"" + url.QueryEscape(value) + "\"")
	}
	result += strings.Join(values, ",")
	return result
}

// Generates the auth signature from the base info and composite key
func generateSig(baseInfo, compositeKey string) string {
	h := hmac.New(sha256.New, []byte(compositeKey))
	h.Write([]byte(baseInfo))
	sha := base64.StdEncoding.EncodeToString(h.Sum(nil))
	return sha 
}

// Generates the base string from the base URL, method, and all of the params
func buildBaseString(baseUrl, method string, params map[string]string) string {
	r := make([]string, 0)
	keys := make([]string, 0)
	for key := range params {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		r = append(r, key + "=" + urlEscape(params[key]))
	}
	return method + "&" + url.QueryEscape(baseUrl) + "&" + url.QueryEscape(strings.Join(r, "&"))
}

// URL encodes the string
func urlEscape(str string) string {
	escaped := url.QueryEscape(str)
	escaped = strings.Replace(escaped, "+", "%20", -1)
	return escaped
}

// Merges the two maps of params into one map of params
func mergeParams(urlParams, oauth map[string]string) map[string]string {
	result := make(map[string]string)
	for key, value := range urlParams {
		result[key] = value
	}

	for key, value := range oauth {
		result[key] = value
	}
	return result;
}

// Converts the params in the url string to a map
func paramsToMap(urlParams string) map[string]string {
	result := make(map[string]string)
	params := strings.Split(urlParams, "&")

	for _, value := range params {
		value, _ := url.PathUnescape(value)
		split := strings.Split(value, "=")
		if len(split) == 2 {
			result[split[0]] = split[1]
		} else {
			result["filter"] = value[7:]
		}
	}
	return result
}

// Generates a random string for the nonce of a given length
func generateNonce(length int) string {
	rand.Seed(time.Now().UnixNano())
	characters := "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
	characterLen := len(characters)
	randomString := ""
	for i := 0; i < length; i++ {
		randomString += string(characters[rand.Intn(characterLen - 1)]);
	}
	return randomString
}